stdlib/array.ml

```diff
@@ -442,8 +442,9 @@ let fast_sort = stable_sort
 let shuffle ~rand a = (* Fisher-Yates *)
   for i = length a - 1 downto 1 do
     let j = rand (i + 1) in
+    if not (0 <= j && j <= i) then invalid_arg "Array.shuffle";
```

Comment:
Thanks @gasche and @dbuenzli for your suggestions. I have improved the error message (without depending on `Printf`) and have placed it in a separate function.

Reply:
Rather than using `Printf` or `^`, I would propose to use  `String.concat`:
```ocaml
    String.concat "" [
      "Array.shuffle: rand("; string_of_int (i + 1); ")"; " returned "; string_of_int j;
      ", outside of the expected range [0; "; string_of_int i; "]"
   ]
 ```
(hm, that's one case where I would have used `[: ... :]` if available)